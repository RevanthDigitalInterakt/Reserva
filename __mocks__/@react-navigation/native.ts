const navigate = jest.fn();

const useNavigation = () => ({
  navigate,
});

export {
  useNavigation,
};
