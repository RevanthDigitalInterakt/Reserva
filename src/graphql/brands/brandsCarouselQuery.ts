import { gql } from '@apollo/client';

export interface IBrandCarouselItem {
  __typename: string,
  brandLogo: {
    __typename: string,
    url: string
  },
  reference: string
}

export interface IBrandsCarouselQuery {
  brandsCarousel: {
    __typename: string,
    brandsCollection: {
      __typename: string,
      items: Array<IBrandCarouselItem>
    }
  }
}

export const brandsCarouselQuery = gql`
    query BrandsCarousel($id: String!){
      brandsCarousel(id: $id) {
        brandsCollection {
          items {
            brandLogo {
              url
            }
            reference
          }
        }
      }
    }
`;
