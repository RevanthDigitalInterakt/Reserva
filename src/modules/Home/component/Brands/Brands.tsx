import React, { useCallback, useEffect, useState } from 'react';
import { Image } from '@usereservaapp/reserva-ui';
import { FlatList, useWindowDimensions } from 'react-native';
import { useLazyQuery } from '@apollo/client';
import { useNavigation } from '@react-navigation/native';
import Config from 'react-native-config';
import { brandsCarouselQuery, IBrandCarouselItem, IBrandsCarouselQuery } from '../../../../graphql/brands/brandsCarouselQuery';
import { BrandContainer, brandShadowContainer, styles } from './styles/styles';
import EventProvider from '../../../../utils/EventProvider';

const BrandsComponent = (): JSX.Element => {
  const [brands, setBrands] = useState<IBrandCarouselItem[]>([]);
  const { width } = useWindowDimensions();
  const navigation = useNavigation();

  const [getBrands] = useLazyQuery<IBrandsCarouselQuery>(brandsCarouselQuery, {
    context: { clientName: 'contentful' },
    variables: {
      id: Config.ID_BRANDS_COLLECTION_CONTENTFUL,
    },
  });

  const handleNavigateToBrand = useCallback((reference: string): void => {
    try {
      if (!reference) {
        navigation.navigate('ProductCatalog');
      }

      navigation.navigate('ProductCatalog', {
        referenceId: reference,
      });
    } catch (error) {
      EventProvider.captureException(error);
    }
  }, []);

  useEffect(() => {
    (async () => {
      try {
        const { data } = await getBrands();

        const items = data?.brandsCarousel?.brandsCollection?.items;

        if (items?.length) {
          setBrands(items);
        }
      } catch (error) {
        EventProvider.captureException(error);
      }
    })();
  }, []);

  return (
    <FlatList
      testID="brands_flatList"
      horizontal
      showsHorizontalScrollIndicator={false}
      data={brands}
      scrollEnabled={brands.length > 4}
      snapToAlignment="start"
      style={styles.carousel}
      keyExtractor={(item) => `idx-brands-${item.brandLogo.url}`}
      contentContainerStyle={styles.contentContainerCarousel}
      scrollEventThrottle={16}
      decelerationRate="fast"
      renderItem={({ item: { brandLogo: { url }, reference }, index }): JSX.Element => (
        <BrandContainer
          testID="brands_brand_container"
          deviceWidth={width}
          index={index}
          onPress={() => handleNavigateToBrand(reference)}
          lastIndex={brands.length}
          style={brandShadowContainer}
        >
          {url ? (
            <Image
              source={{ uri: url }}
              autoHeight
              resizeMode="contain"
              style={styles.brandImageCarousel}
            />
          ) : null}
        </BrandContainer>
      )}
    />

  );
};

export default BrandsComponent;
