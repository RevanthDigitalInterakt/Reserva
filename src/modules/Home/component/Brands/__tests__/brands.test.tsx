import { ThemeProvider } from 'styled-components/native';
import { theme } from '@usereservaapp/reserva-ui';
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react-native';
import { MockedProvider } from '@apollo/client/testing';
import Config from 'react-native-config';
import type { DocumentNode } from 'graphql';
import { act } from 'react-test-renderer';
import Brands from '../Brands';
import { brandsCarouselQuery, IBrandsCarouselQuery } from '../../../../../graphql/brands/brandsCarouselQuery';

interface IApolloMock<T> {
  request: {
    query: DocumentNode,
    variables: object,
  },
  result: {
    data: T
  }
}

const apolloMocks: Array<IApolloMock<IBrandsCarouselQuery>> = [
  {
    request: {
      query: brandsCarouselQuery,
      variables: {
        id: Config.ID_BRANDS_COLLECTION_CONTENTFUL,
      },
    },
    result: {
      data: {
        brandsCarousel: {
          __typename: '_',
          brandsCollection: {
            __typename: '_',
            items: [
              {
                __typename: '_',
                brandLogo: {
                  __typename: '_',
                  url: 'https://teste.com.br',
                },
                reference: 'colecation:188',
              },
            ],
          },
        },
      },
    },
  },
];

const TestingComponent = (
  <ThemeProvider theme={theme}>
    <MockedProvider mocks={apolloMocks} addTypename={false}>
      <Brands />
    </MockedProvider>
  </ThemeProvider>
);

const mockedFn = jest.fn();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: mockedFn }),
}));

describe('Brands tests', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
  });

  describe('Testing brands', () => {
    it.skip('renders without error and match snapshot', () => {
      render(TestingComponent);
      expect(screen.toJSON()).toMatchSnapshot();
    });

    it.skip('check render flatList', async () => {
      render(TestingComponent);

      act(() => { jest.runAllTimers(); });
      await act(async () => {});

      act(() => { jest.runAllTimers(); });
      await act(async () => {});

      expect(screen.queryByTestId('com.usereserva:id/brands_flatList')).toBeVisible();
    });

    it.skip('check brand click', async () => {
      render(TestingComponent);

      act(() => { jest.runAllTimers(); });
      await act(async () => {});

      act(() => { jest.runAllTimers(); });
      await act(async () => {});

      await act(async () => {
        await fireEvent.press(screen.getByTestId('com.usereserva:id/brands_brand_container'));
      });

      expect(mockedFn).toHaveBeenCalledTimes(1);
    });
  });
});
