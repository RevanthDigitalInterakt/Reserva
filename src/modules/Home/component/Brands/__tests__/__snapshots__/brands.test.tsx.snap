// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Brands tests Testing brands renders without error and match snapshot 1`] = `null`;
